src/tz_macos.rs

```diff
@@ -1,80 +1,135 @@
+    let name = if let Some(name) = name.as_utf8() {
+        name
+    } else {
+        // Otherwise convert the name to UTF-8.
+        name.to_utf8(&mut buf)?
     };
-    if CFStringGetBytes(
-        name,
-        range,
-        kCFStringEncodingUTF8,
-        b'\0',
-        false as Boolean,
-        buf.as_mut_ptr(),
-        buf.len() as isize,
-        &mut buf_bytes,
-    ) != range.length
-    {
-        // Could not convert the name.
-        None
-    } else if buf_bytes < 1 || buf_bytes >= MAX_LEN as isize {
+
+    if name.is_empty() || name.len() >= MAX_LEN {
         // The name should not be empty, or excessively long.
         None
     } else {
-        // Convert the name to a `String`.
-        let name = core::str::from_utf8(&buf[..buf_bytes as usize]).ok()?;
         Some(name.to_owned())
     }
 }
 
-struct Dropping<T>(*const T);
+mod system_time_zone {
+    //! create a safe wrapper around `CFTimeZoneRef`
+
+    use core_foundation_sys::base::{CFRelease, CFTypeRef};
+    use core_foundation_sys::timezone::{CFTimeZoneCopySystem, CFTimeZoneGetName, CFTimeZoneRef};
+
+    pub(crate) struct SystemTimeZone(CFTimeZoneRef);
 
-impl<T> Drop for Dropping<T> {
-    #[inline]
-    fn drop(&mut self) {
-        unsafe { CFRelease(self.0 as CFTypeRef) };
+    impl Drop for SystemTimeZone {
+        fn drop(&mut self) {
+            // SAFETY: `SystemTimeZone` is only ever created with a valid `CFTimeZoneRef`.
+            unsafe { CFRelease(self.0 as CFTypeRef) };
+        }
+    }
+
+    impl SystemTimeZone {
+        pub(crate) fn new() -> Option<Self> {
+            // SAFETY: No invariants to uphold. We'll release the pointer when we don't need it anymore.
+            let v: CFTimeZoneRef = unsafe { CFTimeZoneCopySystem() };
+            if v.is_null() {
+                None
+            } else {
+                Some(SystemTimeZone(v))
+            }
+        }
+
+        pub(crate) fn name(&self) -> Option<super::string_ref::StringRef<Self>> {
+            // SAFETY: `SystemTimeZone` is only ever created with a valid `CFTimeZoneRef`.
+            let string = unsafe { CFTimeZoneGetName(self.0) };
+            if string.is_null() {
+                None
+            } else {
+                Some(super::string_ref::StringRef::new(string, self))
+            }
+        }
     }
 }
 
-impl<T> Dropping<T> {
-    #[inline]
-    unsafe fn new(v: *const T) -> Option<Self> {
-        if v.is_null() {
+mod string_ref {
+    //! create safe wrapper around `CFStringRef`
+
+    use std::convert::TryInto;
+
+    use core_foundation_sys::base::{Boolean, CFRange};
+    use core_foundation_sys::string::{
+        kCFStringEncodingUTF8, CFStringGetBytes, CFStringGetCStringPtr, CFStringGetLength,
+        CFStringRef,
+    };
+
+    pub(crate) struct StringRef<'a, T> {
+        string: CFStringRef,
+        _parent: &'a T,
+    }
+
+    impl<'a, T> StringRef<'a, T> {
+        pub(crate) fn new(string: CFStringRef, _parent: &'a T) -> Self {
```

Comment:
This function is unsafe as it requires the passed in pointer to be valid.

Reply:
What if you pass `1 as CFStringRef`? Or `let foo = 0u8; &FOO as *const u8 as CFStringRef`? Both are not valid `CFStringRef`, yet can't be catched with a null pointer check or any check that doesn't involve something like valgrind for that matter.

Reply:
I marked the function as unsafe in 9fbf7a6d7aadc4840181b72b3c4997099be48008